halo2_proofs/src/plonk/evaluation.rs

```diff
@@ -0,0 +1,760 @@
+                    fixed_values,
+                    advice_values,
+                    instance_values,
+                );
+                a * theta + b
+            }
+            Calculation::AddGamma(v) => {
+                v.evaluate(
+                    rotations,
+                    constants,
+                    intermediates,
+                    fixed_values,
+                    advice_values,
+                    instance_values,
+                ) + gamma
+            }
+            Calculation::Store(v) => v.evaluate(
+                rotations,
+                constants,
+                intermediates,
+                fixed_values,
+                advice_values,
+                instance_values,
+            ),
+        }
+    }
+}
+
+/// EvaluationData
+#[derive(Default, Debug)]
+pub struct Evaluator<C: CurveAffine> {
+    /// Constants
+    pub constants: Vec<C::ScalarExt>,
+    /// Rotations
+    pub rotations: Vec<i32>,
+    /// Calculations
+    pub calculations: Vec<CalculationInfo>,
+    /// Value parts
+    pub value_parts: Vec<ValueSource>,
+    /// Lookup results
+    pub lookup_results: Vec<Calculation>,
+}
+
+/// CaluclationInfo
+#[derive(Debug)]
+pub struct CalculationInfo {
+    /// Calculation
+    pub calculation: Calculation,
+    /// How many times this calculation is used
+    pub counter: usize,
+}
+
+impl<C: CurveAffine> Evaluator<C> {
+    /// Creates a new evaluation structure
+    pub fn new(cs: &ConstraintSystem<C::ScalarExt>) -> Self {
+        let mut ev = Evaluator::default();
+        ev.add_constant(&C::ScalarExt::zero());
+        ev.add_constant(&C::ScalarExt::one());
+
+        // Custom gates
+        for gate in cs.gates.iter() {
+            for poly in gate.polynomials().iter() {
+                let vs = ev.add_expression(poly);
+                ev.value_parts.push(vs);
+            }
+        }
+
+        // Lookups
+        for lookup in cs.lookups.iter() {
+            let evaluate_lc = |ev: &mut Evaluator<_>, expressions: &Vec<Expression<_>>| {
+                let parts = expressions
+                    .iter()
+                    .map(|expr| ev.add_expression(expr))
+                    .collect::<Vec<_>>();
+                let mut lc = parts[0];
+                for part in parts.iter().skip(1) {
+                    lc = ev.add_calculation(Calculation::LcTheta(lc, *part));
+                }
+                lc
+            };
+            // Input coset
+            let compressed_input_coset = evaluate_lc(&mut ev, &lookup.input_expressions);
+            // table coset
+            let compressed_table_coset = evaluate_lc(&mut ev, &lookup.table_expressions);
+            // z(\omega X) (a'(X) + \beta) (s'(X) + \gamma)
+            let right_gamma = ev.add_calculation(Calculation::AddGamma(compressed_table_coset));
+            ev.lookup_results
+                .push(Calculation::LcBeta(compressed_input_coset, right_gamma));
+        }
```

Comment:
Currently the gate expressions and lookup input/table expressions are all added together so intermediate results can be shared between all of them. The upside is of course this can save a lot of calculations (with the current EVM circuit this is definitely the case). The downside is that we have to store these values somewhere (`lookup_values`) until they can actually be used in the lookup.

The other option is not to do this for the input/table values of lookups and just calculate those independently on the fly. Calculations may be duplicated, but memory usage is reduced.

With the current EVM circuit the current approach here makes sense I think. With the EVM circuit after https://github.com/appliedzkp/zkevm-circuits/pull/425 maybe less so because the expressions will be fairly simply. On the other hand the number of lookups will be much reduced and the extended domain is also 4x smaller so this will take significantly less memory any way so maybe won't be that much of a deal either way. 

It's really circuit dependent what makes the most sense and ideally the evaluator could even decide itself (perhaps with an optional hint) what approach to take. I was planning on seeing how the circuits evolved to see if this would be worth doing. Does that make sense to you?

Reply:
Yes, totally make sense, I can imagine the alternative approach now, and agree that it's really circuit dependent, perhaps in the future there are different profiles of prover (fast, memory-constraint) for user to decide. 

Thanks for explain this in detailed!